src/mds/MDSRank.cc

```diff
@@ -901,9 +901,29 @@ Session *MDSRank::get_session(Message *m)
 {
   Session *session = static_cast<Session *>(m->get_connection()->get_priv());
   if (session) {
+    session->put(); // do not carry ref
     dout(20) << "get_session have " << session << " " << session->info.inst
 	     << " state " << session->get_state_name() << dendl;
-    session->put();  // not carry ref
```

Comment:
Okay, so obviously the session ref counting is funny since we're returning the pointer without holding a refcount anyway. So I presume we know that's safe anyway, and making use of the session pointer aftering putting a ref count isn't actually racy here, despite the normal rules?

Reply:
Connection::get_priv it does a ->get() so we need to drop the unneded reference. I'm planning to fix this in a future cleanup by assigning the reference to a `boost::intrusive_ptr`.